File: app/src/main/java/com/ternaryop/photoshelf/fragment/ImagePickerFragment.java
package com.ternaryop.photoshelf.fragment;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import android.app.AlertDialog;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.ternaryop.photoshelf.Constants;
import com.ternaryop.photoshelf.ImageUrlRetriever;
import com.ternaryop.photoshelf.R;
import com.ternaryop.photoshelf.activity.ImageViewerActivity;
import com.ternaryop.photoshelf.adapter.ImagePickerAdapter;
import com.ternaryop.photoshelf.adapter.OnPhotoBrowseClickMultiChoice;
import com.ternaryop.photoshelf.adapter.Selection;
import com.ternaryop.photoshelf.dialogs.TumblrPostDialog;
import com.ternaryop.photoshelf.extractor.ImageGallery;
import com.ternaryop.photoshelf.extractor.ImageInfo;
import com.ternaryop.photoshelf.parsers.AndroidTitleParserConfig;
import com.ternaryop.photoshelf.parsers.TitleData;
import com.ternaryop.photoshelf.parsers.TitleParser;
import com.ternaryop.photoshelf.view.AutofitGridLayoutManager;
import com.ternaryop.utils.DialogUtils;
import com.ternaryop.widget.ProgressHighlightViewLayout;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

public class ImagePickerFragment extends AbsPhotoShelfFragment implements OnPhotoBrowseClickMultiChoice, ActionMode.Callback {
    private RecyclerView gridView;
    private ProgressHighlightViewLayout progressHighlightViewLayout;

    private ImageUrlRetriever imageUrlRetriever;
    private ImagePickerAdapter imagePickerAdapter;
    private String detailsText;
    private String parsableTitle;
    ActionMode actionMode;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_image_picker, container, false);
        getActivity().setTitle(R.string.image_picker_activity_title);

        progressHighlightViewLayout = (ProgressHighlightViewLayout) rootView.findViewById(android.R.id.empty);
        progressHighlightViewLayout.setAnimation(AnimationUtils.loadAnimation(getActivity(), R.anim.fade_loop));

        imagePickerAdapter = new ImagePickerAdapter(getActivity());
        imagePickerAdapter.setOnPhotoBrowseClick(this);
        imagePickerAdapter.setEmptyView(progressHighlightViewLayout);
        imageUrlRetriever = new ImageUrlRetriever(getActivity(), (ProgressBar) rootView.findViewById(R.id.progressbar));

        RecyclerView.LayoutManager layout = new AutofitGridLayoutManager(getActivity(), (int) getResources().getDimension(R.dimen.image_picker_grid_width));
        gridView = (RecyclerView) rootView.findViewById(R.id.gridview);
        gridView.setAdapter(imagePickerAdapter);
        gridView.setHasFixedSize(true);
        gridView.setLayoutManager(layout);

        setHasOptionsMenu(true);

        return rootView;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        openUrl(getTextWithUrl());
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        compositeDisposable.clear();
    }

    private String getTextWithUrl() {
        // Search on fragment arguments
        Bundle arguments = getArguments();
        if (arguments != null && arguments.containsKey(Constants.EXTRA_URL)) {
            return arguments.getString(Constants.EXTRA_URL);
        }

        // Search on activity intent
        // Get intent, action and MIME type
        Intent intent = getActivity().getIntent();
        String action = intent.getAction();
        String type = intent.getType();
        Uri uri = intent.getData();

        String textWithUrl = null;

        if (Intent.ACTION_SEND.equals(action) && type != null) {
            if ("text/plain".equals(type)) {
                textWithUrl = intent.getStringExtra(Intent.EXTRA_TEXT);
            }
        } else if (Intent.ACTION_VIEW.equals(action) && uri != null) {
            textWithUrl = uri.toString();
        }
        return textWithUrl;
    }

    private void openUrl(String textWithUrl) {
        if (textWithUrl == null) {
            return;
        }
        progressHighlightViewLayout.startProgress();
        final String message = getResources().getQuantityString(R.plurals.download_url_with_count, 1, 0);
        getCurrentTextView().setText(message);
        final Matcher m = Pattern.compile("(http:.*)").matcher(textWithUrl);

        if (m.find()) {
            readImageGallery(m.group(1));
        } else {
            new AlertDialog.Builder(getActivity())
                    .setTitle(R.string.url_not_found)
                    .setMessage(getString(R.string.url_not_found_description, textWithUrl))
                    .show();
        }
    }

    private void readImageGallery(String url) {
        imageUrlRetriever.readImageGallery(url)
                .doOnSubscribe(new Consumer<Disposable>() {
                    @Override
                    public void accept(Disposable disposable) throws Exception {
                        compositeDisposable.add(disposable);
                    }
                })
                .subscribe(new Consumer<ImageGallery>() {
                               @Override
                               public void accept(ImageGallery gallery) throws Exception {
                                   onGalleryRetrieved(gallery);
                               }
                           },
                new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        DialogUtils.showErrorDialog(getActivity(), throwable);
                    }
                });
    }

    public TextView getCurrentTextView() {
        return (TextView) progressHighlightViewLayout.getCurrentView();
    }

    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
        mode.setTitle(getString(R.string.select_images));
        mode.setSubtitle(getResources().getQuantityString(
                R.plurals.selected_items_total,
                1,
                1,
                imagePickerAdapter.getItemCount()));
        MenuInflater inflater = mode.getMenuInflater();
        inflater.inflate(R.menu.image_picker_context, menu);
        imagePickerAdapter.setShowButtons(true);
        imagePickerAdapter.notifyDataSetChanged();
        return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
        return true;
    }

    @Override
    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
        switch (item.getItemId()) {
            case R.id.showDialog:
                retrieveImages(false);
                mode.finish();
                return true;
            case R.id.create_from_file:
                retrieveImages(true);
                mode.finish();
                return true;
            default:
                return false;
        }
    }

    @Override
    public void onDestroyActionMode(ActionMode mode) {
        this.actionMode = null;
        imagePickerAdapter.setShowButtons(false);
        imagePickerAdapter.getSelection().clear();
    }

    private void retrieveImages(boolean useFile) {
        imageUrlRetriever.retrieve(imagePickerAdapter.getSelectedItems(), useFile)
                .doOnSubscribe(new Consumer<Disposable>() {
                    @Override
                    public void accept(Disposable disposable) throws Exception {
                        compositeDisposable.add(disposable);
                    }
                })
                .toList()
                .subscribe(new Consumer<List<Uri>>() {
                               @Override
                               public void accept(List<Uri> uris) throws Exception {
                                   onImagesRetrieved(uris);
                               }
                           },
                        new Consumer<Throwable>() {
                            @Override
                            public void accept(Throwable throwable) throws Exception {
                                DialogUtils.showSimpleMessageDialog(getActivity(), R.string.url_not_found, throwable.getLocalizedMessage());
                            }
                        });
    }

    public void onImagesRetrieved(List<Uri> imageUriList) {
        try {
            TitleData titleData = TitleParser.instance(new AndroidTitleParserConfig(getActivity())).parseTitle(parsableTitle);
            Bundle args = new Bundle();

            args.putParcelableArrayList(TumblrPostDialog.ARG_IMAGE_URLS, new ArrayList<>(imageUriList));
            args.putString(TumblrPostDialog.ARG_HTML_TITLE, titleData.toHtml());
            args.putString(TumblrPostDialog.ARG_SOURCE_TITLE, parsableTitle);

            args.putStringArrayList(TumblrPostDialog.ARG_INITIAL_TAG_LIST, new ArrayList<>(titleData.getTags()));

            TumblrPostDialog.newInstance(args, null).show(getFragmentManager(), "dialog");
        } catch (Exception e) {
            new AlertDialog.Builder(getActivity())
                    .setTitle(R.string.parsing_error)
                    .setMessage(e.getLocalizedMessage())
                    .show();
        }
    }

    public void onGalleryRetrieved(ImageGallery imageGallery) {
        progressHighlightViewLayout.stopProgress();
        detailsText = imageGallery.getTitle();
        parsableTitle = buildParsableTitle(imageGallery);
        showDetails(Snackbar.LENGTH_LONG);
        final List<ImageInfo> imageInfoList = imageGallery.getImageInfoList();
        getSupportActionBar().setSubtitle(getResources().getQuantityString(R.plurals.image_found, imageInfoList.size(), imageInfoList.size()));
        imagePickerAdapter.addAll(imageInfoList);
    }

    @Override
    public void onTagClick(int position) {
    }

    @Override
    public void onThumbnailImageClick(int position) {
        final ImageInfo imageInfo = imagePickerAdapter.getItem(position);
        if (imageInfo.getImageUrl() == null) {
            List<ImageInfo> imageInfoList = new ArrayList<>();
            imageInfoList.add(imageInfo);
            imageUrlRetriever.retrieve(imageInfoList, false)
                    .doOnSubscribe(new Consumer<Disposable>() {
                        @Override
                        public void accept(Disposable disposable) throws Exception {
                            compositeDisposable.add(disposable);
                        }
                    })
                    .take(1)
                    .subscribe(new Consumer<Uri>() {
                                   @Override
                                   public void accept(Uri uri) throws Exception {
                                       // cache retrieved value
                                       final String url = uri.toString();
                                       imageInfo.setImageUrl(url);
                                       ImageViewerActivity.startImageViewer(getActivity(), url, null);
                                   }
                               },
                            new Consumer<Throwable>() {
                                @Override
                                public void accept(Throwable throwable) throws Exception {
                                    DialogUtils.showSimpleMessageDialog(getActivity(), R.string.url_not_found, throwable.getLocalizedMessage());
                                }
                            });
        } else {
            ImageViewerActivity.startImageViewer(getActivity(), imageInfo.getImageUrl(), null);
        }
    }

    @Override
    public void onOverflowClick(View view, int position) {
    }

    @Override
    public void onItemClick(int position) {
        if (actionMode == null) {
            onThumbnailImageClick(position);
        } else {
            updateSelection(position);
        }
    }

    @Override
    public void onItemLongClick(int position) {
        if (actionMode == null) {
            actionMode = getActivity().startActionMode(this);
        }
        updateSelection(position);
    }

    private void updateSelection(int position) {
        Selection selection = imagePickerAdapter.getSelection();
        selection.toggle(position);
        if (selection.getItemCount() == 0) {
            actionMode.finish();
        } else {
            int selectionCount = selection.getItemCount();
            actionMode.setSubtitle(getResources().getQuantityString(
                    R.plurals.selected_items_total,
                    selectionCount,
                    selectionCount,
                    imagePickerAdapter.getItemCount()));
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.image_picker, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_image_viewer_details:
                showDetails(Snackbar.LENGTH_INDEFINITE);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void showDetails(int duration) {
        Snackbar snackbar = Snackbar.make(gridView, detailsText, duration);
        View sbView = snackbar.getView();
        sbView.setBackgroundColor(ContextCompat.getColor(getActivity(), R.color.image_picker_detail_text_bg));
        TextView textView = (TextView) sbView.findViewById(android.support.design.R.id.snackbar_text);
        textView.setTextColor(ContextCompat.getColor(getActivity(), R.color.image_picker_detail_text_text));
        textView.setMaxLines(3);
        snackbar.show();
    }

    /**
     * Return a string that can be used by the title parser
     * @param imageGallery the source
     * @return the title plus domain string
     */
    public String buildParsableTitle(final ImageGallery imageGallery) {
        return imageGallery.getTitle() + " ::::: " + imageGallery.getDomain();
    }

}

File: app/src/main/java/com/ternaryop/photoshelf/fragment/ImagePickerFragment.kt
package com.ternaryop.photoshelf.fragment

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.ActionMode
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.view.animation.AnticipateOvershootInterpolator
import android.widget.ProgressBar
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.constraintlayout.widget.ConstraintSet
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager.HORIZONTAL
import androidx.recyclerview.widget.RecyclerView
import androidx.transition.ChangeBounds
import androidx.transition.TransitionManager
import com.google.android.material.snackbar.Snackbar
import com.ternaryop.photoshelf.EXTRA_URL
import com.ternaryop.photoshelf.R
import com.ternaryop.photoshelf.activity.ImageViewerActivity
import com.ternaryop.photoshelf.adapter.ImagePickerAdapter
import com.ternaryop.photoshelf.adapter.OnPhotoBrowseClickMultiChoice
import com.ternaryop.photoshelf.api.extractor.ImageGallery
import com.ternaryop.photoshelf.api.extractor.ImageInfo
import com.ternaryop.photoshelf.dialogs.PostDialogData
import com.ternaryop.photoshelf.dialogs.TumblrPostDialog
import com.ternaryop.photoshelf.lifecycle.Status
import com.ternaryop.utils.dialog.DialogUtils
import com.ternaryop.utils.dialog.showErrorDialog
import com.ternaryop.utils.recyclerview.AutofitGridLayoutManager
import com.ternaryop.widget.ProgressHighlightViewLayout

const val MAX_DETAIL_LINES = 3

class ImagePickerFragment : AbsPhotoShelfFragment(), OnPhotoBrowseClickMultiChoice, ActionMode.Callback {
    private lateinit var gridView: RecyclerView
    private lateinit var progressHighlightViewLayout: ProgressHighlightViewLayout
    private lateinit var progressbar: ProgressBar

    private lateinit var imagePickerAdapter: ImagePickerAdapter
    private lateinit var selectedItemsViewContainer: SelectedItemsViewContainer
    private lateinit var imageGallery: ImageGallery
    private lateinit var viewModel: ImagePickerViewModel

    // Search on fragment arguments
    private val textWithUrl: String?
        get() {
            val arguments = arguments
            if (arguments != null && arguments.containsKey(EXTRA_URL)) {
                return arguments.getString(EXTRA_URL)
            }
            // Search on activity intent
            // Get intent, action and MIME type
            val intent = requireActivity().intent
            val action = intent.action
            val type = intent.type
            val uri = intent.data

            var textWithUrl: String? = null

            if (Intent.ACTION_SEND == action && type != null) {
                if ("text/plain" == type) {
                    textWithUrl = intent.getStringExtra(Intent.EXTRA_TEXT)
                }
            } else if (Intent.ACTION_VIEW == action && uri != null) {
                textWithUrl = uri.toString()
            }
            return textWithUrl
        }

    private val currentTextView: TextView
        get() = progressHighlightViewLayout.currentView as TextView

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_image_picker, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewModel = ViewModelProviders.of(this).get(ImagePickerViewModel::class.java)

        viewModel.result.observe(this, Observer { result ->
            when (result) {
                is ImagePickerModelResult.Gallery -> onGalleryModelResult(result)
                is ImagePickerModelResult.ImageList -> onImageListModelResult(result)
                is ImagePickerModelResult.Image -> onImageModelResult(result)
            }
        })

        setHasOptionsMenu(true)
        setupUI(view, requireContext())
    }

    private fun onGalleryModelResult(result: ImagePickerModelResult.Gallery) {
        when (result.command.status) {
            Status.SUCCESS -> onGalleryRetrieved(result.command.data?.gallery)
            Status.ERROR -> showError(result.command.error, false)
            Status.PROGRESS -> { }
        }
    }

    private fun onImageListModelResult(result: ImagePickerModelResult.ImageList) {
        when (result.command.status) {
            Status.SUCCESS -> onRetrievedImageList(result.command.data?.map { it.second })
            Status.ERROR -> showError(result.command.error, true)
            Status.PROGRESS -> progressbar.incrementProgressBy(1)
        }
    }

    private fun onImageModelResult(result: ImagePickerModelResult.Image) {
        when (result.command.status) {
            Status.SUCCESS -> onRetrievedSingleImage(result.command.data)
            Status.ERROR -> showError(result.command.error, true)
            Status.PROGRESS -> progressbar.incrementProgressBy(1)
        }
    }

    private fun showError(error: Throwable?, showAlert: Boolean) {
        progressbar.visibility = GONE
        if (showAlert) {
            DialogUtils.showSimpleMessageDialog(requireContext(), R.string.url_not_found, error?.localizedMessage ?: "")
        } else {
            showSnackbar(makeSnack(gridView, error?.localizedMessage ?: "") { refreshUI() })
        }
    }

    private fun onRetrievedSingleImage(uri: ImageInfoUriPair?) {
        progressbar.visibility = GONE
        uri ?: return
        // cache retrieved value
        val url = uri.second.toString()
        uri.first.imageUrl = url
        ImageViewerActivity.startImageViewer(requireContext(), url)
    }

    private fun setupUI(view: View, context: Context) {
        progressHighlightViewLayout = view.findViewById(android.R.id.empty)
        progressHighlightViewLayout.progressAnimation = AnimationUtils.loadAnimation(context, R.anim.fade_loop)
        progressHighlightViewLayout.visibility = View.VISIBLE

        imagePickerAdapter = ImagePickerAdapter(context)
        imagePickerAdapter.setOnPhotoBrowseClick(this)
        imagePickerAdapter.setEmptyView(progressHighlightViewLayout)

        val layoutManager = AutofitGridLayoutManager(context,
            resources.getDimension(R.dimen.image_picker_grid_width).toInt())
        gridView = view.findViewById(R.id.gridview)
        gridView.adapter = imagePickerAdapter
        gridView.setHasFixedSize(true)
        gridView.layoutManager = layoutManager
        // animating the constraintLayout results in grid animations, so we disable them
        gridView.itemAnimator = null

        selectedItemsViewContainer = SelectedItemsViewContainer(
            context,
            view.findViewById(R.id.constraintlayout),
            view.findViewById(R.id.selectedItems))
        selectedItemsViewContainer.adapter
            .setOnPhotoBrowseClick(object: OnPhotoBrowseClickMultiChoice {
                override fun onItemClick(position: Int) {
                    val index = imagePickerAdapter.getIndex(selectedItemsViewContainer.adapter.getItem(position))
                    layoutManager.scrollToPosition(index)
                }

                override fun onItemLongClick(position: Int) {}
                override fun onTagClick(position: Int, clickedTag: String) {}
                override fun onThumbnailImageClick(position: Int) {}
                override fun onOverflowClick(position: Int, view: View) {}
            })

        progressbar = view.findViewById(R.id.progressbar)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        requireActivity().setTitle(R.string.image_picker_activity_title)

        refreshUI()
    }

    override fun refreshUI() = openUrl(textWithUrl)

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        job.cancel()
    }

    private fun openUrl(textWithUrl: String?) {
        if (textWithUrl == null) {
            return
        }
        progressHighlightViewLayout.startProgress()
        val message = resources.getQuantityString(R.plurals.download_url_with_count, 1, 0)
        currentTextView.text = message
        val url = "(https?:.*)".toRegex().find(textWithUrl)?.value

        if (url == null) {
            AlertDialog.Builder(requireContext())
                .setTitle(R.string.url_not_found)
                .setMessage(getString(R.string.url_not_found_description, textWithUrl))
                .show()
        } else {
            viewModel.readImageGallery(url)
        }
    }

    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
        mode.title = getString(R.string.select_images)
        mode.subtitle = resources.getQuantityString(
                R.plurals.selected_items_total,
                1,
                1,
                imagePickerAdapter.itemCount)
        mode.menuInflater.inflate(R.menu.image_picker_context, menu)
        imagePickerAdapter.showButtons = true
        imagePickerAdapter.notifyDataSetChanged()
        return true
    }

    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
        return true
    }

    override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.showDialog -> {
                showProgressbar(imagePickerAdapter.selectedItems.size)
                viewModel.imageList(imagePickerAdapter.selectedItems)
                finish(mode)
                true
            }
            R.id.create_from_file -> {
                showProgressbar(imagePickerAdapter.selectedItems.size)
                viewModel.imageList(imagePickerAdapter.selectedItems, requireContext().cacheDir)
                finish(mode)
                true
            }
            R.id.show_selected_items -> {
                selectedItemsViewContainer.toggleVisibility()
                true
            }
            else -> false
        }
    }

    override fun onDestroyActionMode(mode: ActionMode) {
        this.actionMode = null
        imagePickerAdapter.showButtons = false
        imagePickerAdapter.selection.clear()
        selectedItemsViewContainer.updateList(emptyList())
    }

    private fun finish(mode: ActionMode) {
        mode.finish()
        // if visibility is set (to GONE) on layout file the constraintLayout restores it to GONE after any view modification
        // this means the progress bar is not visible while progress goes on
        // The workaround consists to make the progressbar visible when the actionMode is 'finished' programmatically
        progressbar.visibility = VISIBLE
    }

    private fun onRetrievedImageList(imageUriList: List<Uri>?) {
        progressbar.visibility = GONE

        imageUriList ?: return
        try {
            fragmentManager?.also {
                TumblrPostDialog.newInstance(PostDialogData(imageGallery.parsableTitle,
                    imageGallery.titleParsed.html, imageGallery.titleParsed.tags, imageUriList), null)
                    .show(it, "dialog")
            }
        } catch (e: Exception) {
            e.showErrorDialog(requireContext())
        }
    }

    private fun onGalleryRetrieved(imageGallery: ImageGallery?) {
        progressHighlightViewLayout.stopProgress()

        imageGallery ?: return

        this.imageGallery = imageGallery
        showDetails(Snackbar.LENGTH_LONG)
        val imageInfoList = imageGallery.imageInfoList
        supportActionBar?.subtitle = resources.getQuantityString(R.plurals.image_found,
            imageInfoList.size, imageInfoList.size)
        imagePickerAdapter.addAll(imageInfoList)
    }

    override fun onTagClick(position: Int, clickedTag: String) {}

    override fun onThumbnailImageClick(position: Int) {
        val imageInfo = imagePickerAdapter.getItem(position)
        val imageUrl = imageInfo.imageUrl
        if (imageUrl == null) {
            showProgressbar(1)
            viewModel.image(imageInfo)
        } else {
            ImageViewerActivity.startImageViewer(requireContext(), imageUrl)
        }
    }

    override fun onOverflowClick(position: Int, view: View) {}

    override fun onItemClick(position: Int) {
        if (actionMode == null) {
            onThumbnailImageClick(position)
        } else {
            updateSelection(position)
        }
    }

    override fun onItemLongClick(position: Int) {
        if (actionMode == null) {
            actionMode = requireActivity().startActionMode(this)
        }
        updateSelection(position)
    }

    private fun updateSelection(position: Int) {
        val selection = imagePickerAdapter.selection
        selection.toggle(position)
        selectedItemsViewContainer.updateList(imagePickerAdapter.selectedItems)
        if (selection.itemCount == 0) {
            actionMode?.finish()
        } else {
            val selectionCount = selection.itemCount
            actionMode?.subtitle = resources.getQuantityString(
                    R.plurals.selected_items_total,
                    selectionCount,
                    selectionCount,
                    imagePickerAdapter.itemCount)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.image_picker, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_image_viewer_details -> {
                showDetails(Snackbar.LENGTH_INDEFINITE)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun showDetails(duration: Int) {
        val snackbar = Snackbar.make(gridView, imageGallery.title ?: "No title", duration)
        val sbView = snackbar.view
        sbView.setBackgroundColor(ContextCompat.getColor(requireContext(), R.color.image_picker_detail_text_bg))
        val textView = sbView.findViewById<TextView>(com.google.android.material.R.id.snackbar_text)
        textView.setTextColor(ContextCompat.getColor(requireContext(), R.color.image_picker_detail_text_text))
        textView.maxLines = MAX_DETAIL_LINES
        snackbar.show()
    }

    private fun showProgressbar(max: Int) {
        progressbar.progress = 0
        progressbar.max = max
        progressbar.visibility = VISIBLE
    }
}

private class SelectedItemsViewContainer(
    val context: Context,
    private val constraintLayout: ConstraintLayout,
    selectionListView: RecyclerView) {
    val adapter = ImagePickerAdapter(context)
    private var isVisible = false
    private val constraintSet = ConstraintSet()

    init {
        selectionListView.adapter = adapter
        selectionListView.setHasFixedSize(true)
        selectionListView.layoutManager = LinearLayoutManager(context, HORIZONTAL, false)

        constraintSet.clone(context, R.layout.fragment_image_picker)
    }

    fun toggleVisibility() = show(!isVisible)

    fun show(show: Boolean) {
        isVisible = show
        if (isVisible) {
            constraintSet.clear(R.id.selectedItems, ConstraintSet.TOP)
        } else {
            constraintSet.connect(R.id.selectedItems, ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.BOTTOM)
        }

        val transition = ChangeBounds()
        transition.interpolator = AnticipateOvershootInterpolator(1.0f)
        transition.duration = 1000

        TransitionManager.beginDelayedTransition(constraintLayout, transition)

        constraintSet.applyTo(constraintLayout)
    }

    fun updateList(items: List<ImageInfo>) {
        adapter.clear()
        if (items.isEmpty()) {
            show(false)
        } else {
            adapter.addAll(items)
        }
    }
}
